compsequential.go

```diff
@@ -0,0 +1,210 @@
+package routinghelpers
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/ipfs/go-cid"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/routing"
+)
+
+var _ routing.Routing = &ComposableSequential{}
+
+type ComposableSequential struct {
+	routers []*SequentialRouter
+}
+
+func NewComposableSequential(routers []*SequentialRouter) *ComposableSequential {
+	return &ComposableSequential{
+		routers: routers,
+	}
+}
+
+// Provide calls Provide method per each router sequentiall.
+// If some router fails and the IgnoreError flag is true, we continue to the next router.
+// Context timeout error will be also ignored if the flag is set.
+func (r *ComposableSequential) Provide(ctx context.Context, cid cid.Cid, provide bool) error {
+	for _, router := range r.routers {
+		ctx, cancel := context.WithTimeout(ctx, router.Timeout)
+		if err := router.Router.Provide(ctx, cid, provide); err != nil &&
+			!errors.Is(err, routing.ErrNotSupported) &&
+			!router.IgnoreError {
+			cancel()
+			return err
+		}
+		cancel()
+	}
+
+	return nil
+}
+
+// FindProvidersAsync calls FindProvidersAsync per each router sequentially.
+// If some router fails and the IgnoreError flag is true, we continue to the next router.
+// Context timeout error will be also ignored if the flag is set.
+// If count is set, the channel will return up to count results, stopping routers iteration.
+func (r *ComposableSequential) FindProvidersAsync(ctx context.Context, cid cid.Cid, count int) <-chan peer.AddrInfo {
+	chanOut := make(chan peer.AddrInfo)
+	sentCount := 0
+	go func() {
+		for _, router := range r.routers {
+			ctx, cancel := context.WithTimeout(ctx, router.Timeout)
+			rch := router.Router.FindProvidersAsync(ctx, cid, count)
+
+			var g sync.WaitGroup
+			g.Add(1)
+			go func() {
+				defer g.Done()
+				for {
+					select {
+					case <-ctx.Done():
+						return
```

Comment:
If the context is cancelled this cancels the current router but then it continues w/ the rest of the routers, we probably ought to short-circuit the whole loop at this point to avoid kicking off a bunch of wasteful queries.

Reply:
hmm, that's tricky, because that context is the specific router context with the timeout, so maybe the next routers still have execution time.